tests from Gromova

Purka opened the pull request:
I'm not sure I'm doing it right it's a bit harder than games

Commit:
added tests

lastperson commented on test/debts.js:
```diff
@@ -32,6 +41,15 @@ contract('Debts', function(accounts) {
+   it('added: should fail on underflow when repaing', () => {
+    const borrower = accounts[3];
+    const value = 100;
+    const underValue = '0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
```
In order for underflow to happen, you just need a number bigger than value itself, say 101. But this also works.

lastperson commented on test/debts.js:
```diff
@@ -47,13 +65,44 @@ contract('Debts', function(accounts) {
+  });
 
-  it('should direct you for inventing more tests');
+  it('added: should not to be overpaid when repaing', () => {
```
pretty much the same covered above.

lastperson reviewed (approved):
What should happen with partial repayment?
What should happen with 0 borrow or 0 repayment?
Any othe cases to cover?

Commit:
added new tests

Purka commented:
I've added some more tests. I couldn't think of more ._.

Commit:
Added Borrow.sol and borrow.js

I added my version of contract and some tests for it

Purka commented:
I've done and ready for feedback

lastperson commented on contracts/Borrow.sol:
```diff
@@ -0,0 +1,107 @@
+        if (_amount == 0) {
+            Error("Zero is not valid");
+            return false;
+        } else if (requested[msg.sender] + _amount < requested[msg.sender]) {
```
If there is a return statement in 'if' then maybe you can use just 'if' instead of 'else if' on the next line? Probably will become easier to read.

lastperson commented on contracts/Borrow.sol:
```diff
@@ -0,0 +1,107 @@
+    //Borrower can cancel his request 
+    function reduceRequestBorrow(uint _amount) onlyNotOwner public returns(bool success) {
+            if (_amount == 0) {
+            Error("Zero is not valid");
```
indentation

lastperson commented on contracts/Borrow.sol:
```diff
@@ -0,0 +1,107 @@
+            if (_amount == 0) {
+            Error("Zero is not valid");
+            return false;
+            } else if (requested[msg.sender] - _amount > requested[msg.sender]) {
```
`if (requested < _amount)` maybe it is simpler this way?

lastperson commented on contracts/Borrow.sol:
```diff
@@ -0,0 +1,107 @@
+            return false;
+        } else if (_amount <= amount) {
+            borrows[_requester] += _amount;
+            requested[_requester] -= amount;
```
requested = 0 ?

lastperson replied:
having the successful outcome somewhere inbetween fail outcomes confuses reader.

lastperson commented on contracts/Borrow.sol:
```diff
@@ -0,0 +1,107 @@
+            Error("it's Underflow");
+            return false;
+        } else {
+            require(borrows[msg.sender] >= _amount);
```
Is there a trst for this check?

Purka replied:
Yes, it('should not be overpaid when repaying')
But actually, this line is redundant. I'll fix it

Commit:
Comments are taken into account

Commit:
OXO game

lastperson commented on contracts/OXO.sol:
```diff
@@ -0,0 +1,235 @@
+	 * Return true only if it's win combo
+	 * Return false in other cases
+	 */
+	function isWinner(byte s) public returns (bool success) {
```
Looks like constant.

lastperson reviewed (changes requested):
1. Is it possible to play another game after the first one?
2. What if players don't withdraw and another game starts?

Commit:
Corrected

added test that checks isWinner
added ability to play more than one game

Commit:
added time stamp

If someone has disappeared during the game then remaining player is winner
Corrected the problem with withdraw

lastperson commented on test/oxotest.js:
```diff
@@ -473,7 +473,46 @@ contract('OXO', function (accounts) {
+		});
+	});
+	
+	it('shoud allow to play again after returnDeposit', () => {
```
not finished?

lastperson commented on contracts/OXO.sol:
```diff
@@ -238,4 +254,23 @@ contract OXO {
+
+	function returnDeposit() public returns (bool success) {
+		if (now >= lastMoveTime + 30 seconds) {
+			if (lastMove == '0') {
```
If there is not tests, then there is most probably a bug.

lastperson reviewed (changes requested):
There is also places of code duplication that can be refactored.

Commit:
tests complited

Added more tests for gerReward function and corrected some bugs there

Commit:
corrected some mistakes
